PXC-3040: Assertion when ALTER TABLE(SPACE) executed concurrently with another query

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3040

Fixed access protection when accessing DD cached objects.
MDL lock was locked, then objects were accessed, their reference count was incremented. Then MDL lock was unlocked before reference count was decreased.

Additional problem was the lack of unlocking of LOCK_wsrep_alter_tablespace if lock_tablespace_names() failed for ALTER TABLESPACE.

kamil-holubicki commented:
https://pxc.cd.percona.com/view/PXC%208.0/job/pxc80-test-mtr/
MTR test failure due to test instability (not related to changes)

dutow commented on sql/sql_admin.cc:
```diff
@@ -1515,20 +1522,16 @@ static bool pxc_strict_mode_admin_check(THD *thd, TABLE_LIST *tables) {
       if (table_ref == nullptr ||
```
There's also a release_lock above this if

kamil-holubicki committed:
PXC-3040: Assertion when ALTER TABLE(SPACE) executed concurrently with another query

https://jira.percona.com/browse/PXC-3040

Fixed access protection when accessing DD cached objects.
MDL lock was locked, then objects were accessed, their reference count was incremented. Then MDL lock was unlocked before reference count was decreased.

Additional problem was the lack of unlocking of LOCK_wsrep_alter_tablespace if lock_tablespace_names() failed for ALTER TABLESPACE.

kamil-holubicki commented:
Addressed PR comments + rebased on trunk: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc80-test-mtr/186/

kennt-percona reviewed (approved):
lgtm
